src/shared/rm-rf.c

```diff
@@ -425,11 +425,14 @@ static int rm_rf_children_impl(
         return ret;
 }
 
-int rm_rf(const char *path, RemoveFlags flags) {
+int rm_rf_at(int dir_fd, const char *path, RemoveFlags flags) {
         mode_t old_mode;
         int fd, r, q = 0;
 
-        assert(path);
+        assert(dir_fd >= 0 || dir_fd == AT_FDCWD);
+
+        if (FLAGS_SET(flags, REMOVE_ROOT))
+                assert(path && !dot_or_dot_dot(path)); /* unlinkat() does not support AT_EMPTY_PATH or "." so a path must be provided here. */
```

Comment:
hmm, i think this should be a runtime failure, since paths often are controlled by users (as opposed to fd args for example).

hence:

```c
if (FLAGS_SET(flags, REMOVE_ROOT) && (!path || dot_or_dot_dot(path))
        return -EINVAL;
```

Reply:
hmm, and what happens if people call `rm_rf_at(some_fd, "../..")`?

the check should be a different one here, no?